osu.Game/Screens/Play/PlayerSettings/BeatmapOffsetControl.cs

```diff
@@ -0,0 +1,212 @@
+                Direction = FillDirection.Vertical,
+                Spacing = new Vector2(10),
+                Children = new Drawable[]
+                {
+                    new PlayerSliderBar<double>
+                    {
+                        KeyboardStep = 5,
+                        LabelText = BeatmapOffsetControlStrings.BeatmapOffset,
+                        Current = Current,
+                    },
+                    referenceScoreContainer = new FillFlowContainer
+                    {
+                        Spacing = new Vector2(10),
+                        Direction = FillDirection.Vertical,
+                        RelativeSizeAxes = Axes.X,
+                        AutoSizeAxes = Axes.Y,
+                    },
+                }
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            ReferenceScore.BindValueChanged(scoreChanged, true);
+
+            beatmapOffsetSubscription = realm.RegisterCustomSubscription(r =>
+            {
+                var userSettings = r.Find<BeatmapInfo>(beatmap.Value.BeatmapInfo.ID)?.UserSettings;
+
+                if (userSettings == null) // only the case for tests.
+                    return null;
+
+                Current.Value = userSettings.Offset;
+                userSettings.PropertyChanged += onUserSettingsOnPropertyChanged;
+
+                return new InvokeOnDisposal(() => userSettings.PropertyChanged -= onUserSettingsOnPropertyChanged);
+
+                void onUserSettingsOnPropertyChanged(object sender, PropertyChangedEventArgs args)
+                {
+                    if (args.PropertyName == nameof(BeatmapUserSettings.Offset))
+                        Current.Value = userSettings.Offset;
+                }
+            });
+
+            Current.BindValueChanged(currentChanged);
+        }
+
+        private Task realmWrite;
+
+        private void currentChanged(ValueChangedEvent<double> offset)
+        {
+            Scheduler.AddOnce(updateOffset);
+
+            void updateOffset()
+            {
+                // ensure the previous write has completed. ignoring performance concerns, if we don't do this, the async writes could be out of sequence.
+                if (realmWrite?.IsCompleted == false)
+                {
+                    Scheduler.AddOnce(updateOffset);
+                    return;
+                }
+
+                if (useAverageButton != null)
+                    useAverageButton.Enabled.Value = !Precision.AlmostEquals(lastPlayAverage, Current.Value, Current.Precision);
+
+                realmWrite = realm.WriteAsync(r =>
+                {
+                    var settings = r.Find<BeatmapInfo>(beatmap.Value.BeatmapInfo.ID)?.UserSettings;
+
+                    if (settings == null) // only the case for tests.
+                        return;
+
+                    if (Precision.AlmostEquals(settings.Offset, Current.Value))
```

Comment:
Precision spec isn't present here compared to the other comparison just a few lines above.

Reply:
I'm not sure about this one. I am thinking of dropping the `Precision` call altogether as what is stored in realm should be a precisely matching `double`. Hopefully that is true across all platforms..

Reply:
Hmm fair call. I reckon it should be safe to drop the precision check, bindable number rounding logic should prevent this from firing too much.

C# generally doesn't give any guarantees wrt floating point behaviour but I'd like to think that if it ever changes it'll be due to a major .NET release at which point this'll run maybe once after the bump and then not run again for a long time.